---
title: "TPM_Normalization"
author: "Cera Fisher"
date: "September 14, 2018"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Scaling TPM between two species with different numbers of annotated transcripts

_based on Musser & Wagner (2015), JEZ:B_

TPM, transcripts per million mapped, is a way of normalizing RNAseq data that accounts for differences in library size by scaling the abundance of a transcript (the "counts") to the total number of transcripts assumed to be present in the transcriptome. In short, TPM = count for transcript "i" / total number of annotated transcripts * a scaling factor. 

Necessarily, TPM from one species does not map to the TPM of another species if their transcriptomes are of different sizes, which they almost certainly are, so the values for the species with the smaller transcriptome will be inflated relative to the species with the larger transcriptome. According to Musser & Wagner, these values can be rescaled by calculating a scaling factor $\alpha$:
 $$\alpha = 10^{-6} \sum_{j=1}^{N1} {tpm(Aj)}$$
 
 Where N1 = the number of transcripts for species B (the smaller set), _j_ = all the transcripts in the set, and tpm(Aj) is the transcripts per million for species A for each of the genes _j_ in the set. 

Here is how I interpret this to work in my transcriptomes for _Entylia carinata_ and _Homalodisca vitripennis_. 


1) Read in the un-normalized TPM values from RSEM/edgeR. 

```{r warning=FALSE, message=FALSE, tidy=TRUE}
library("dplyr")
options(stringsAsFactors = FALSE)
Ecar.TPM <- read.table("C:/Users/cruth/Google Drive/Treehoppers/ResearchFiles/RNASeq/GeneExpression_2018/Refined/EC/ECEF_allSamples.isoform.TPM.not_cross_norm", 
                      sep="\t", header=TRUE)
colnames(Ecar.TPM)
colnames(Ecar.TPM)[1] <- "ECid"

Hvit.TPM <- read.table("C:/Users/cruth/Google Drive/Treehoppers/ResearchFiles/RNASeq/GeneExpression_2018/Refined/HV/CountsMatrices/HV_allSamples.isoform.TPM.not_cross_norm", 
                      sep="\t", header=TRUE)
colnames(Hvit.TPM)
colnames(Hvit.TPM)[1] <- "HVid"

```

2) Calculate $\alpha$ for *H. vitripennis*:

```{r tidy=TRUE}

HvitN1 <- as.numeric(length(Hvit.TPM$HVid))
                                            # 19,126
EcarN2 <- as.numeric(length(Ecar.TPM$ECid))
                                            # 19,975
sumN2.EcarTPM <- sum(Ecar.TPM[,2:42])/41
sumN2.EcarTPM

## Sum of TPM for any given sample should, by definition, be 1,000,000
# The average TPM for Ecar is the sum divided by the number of transcripts. 
Ec.avg.TPM <- sumN2.EcarTPM/19975

## Getting the sum of Ecar TPM for the # of transcripts in Hvit's transcriptome
## i.e, multiplying the average times 19,126
sum.ECavgTPM.HvitN1 <- Ec.avg.TPM * HvitN1

# To get alpha, divide that amount by 1,000,000
alpha <- sum.ECavgTPM.HvitN1 * (10**(-6))

alpha


```
This value for $\alpha$, 0.957....etc is very close to the ratio of the smaller number of transcripts to the larger:

```{r tidy=TRUE}
checksum <- HvitN1/EcarN2
checksum

checksum - alpha
```

Which perhaps should be expected, since those are the only values in this calculation that don't cancel out. 


3) Calculating alpha for *Oncopeltus fasciatus*: 

```{r tidy=TRUE}
Ofas.TPM <- read.csv("C:/Users/cruth/Google Drive/Treehoppers/ResearchFiles/RNASeq/GeneExpression_2018/Refined/OF/Ofas_i5k.genes.isoform.TPM.not_cross_norm.matrix", sep="\t")

names(Ofas.TPM)[1] <- "OFid"
OfasN2 <- length(Ofas.TPM$OFid)

sum.ECavgTPM.OfasN2 <- Ec.avg.TPM * OfasN2

OFalpha <- sum.ECavgTPM.OfasN2 * (10**(-6))

```


### Scaling Hvit.TPM

```{r tidy=TRUE}
HV_scaled <- Hvit.TPM[,-1]
row.names(HV_scaled) <- Hvit.TPM[,1]
Hvit.TPM.scaled <- HV_scaled * alpha
dim(HV_scaled)                       

head(Hvit.TPM[,3])
head(Hvit.TPM.scaled[,2])

```

### Scaling Ofas.TPM 
```{r tidy=TRUE}
OF_scaled <- Ofas.TPM[,-1]
row.names(OF_scaled) <- Ofas.TPM[,1]
Ofas.TPM.scaled <- OF_scaled * OFalpha
dim(OF_scaled)                       

head(Ofas.TPM[,3])
head(Ofas.TPM.scaled[,2])

```

Multiplying by $\alpha$ results in our Hvit TPM numbers being just a little smaller, though it will matter a lot for some of the outrageously large numbers:

```{r tidy=TRUE, collapse=TRUE}
which(Hvit.TPM[,3] == max(Hvit.TPM[,3]))
#8504
Hvit.TPM[,3][8504]

Hvit.TPM.scaled[,2][8504]

which(Ofas.TPM[,3] == max(Ofas.TPM[,3]))

Ofas.TPM[,3][15507]

Ofas.TPM.scaled[,2][15507]
```
Finally, we want to normalize for size using DESeq2, and then write the resulting matrices to file for filtering to single copy orthologs and other downstream analysis. 

```{r eval=FALSE, tidy=TRUE}


write.table(Ofas.TPM.scaled, "C:/Users/cruth/Google Drive/Treehoppers/SRAProject/WGCNA/Ofasciatus/Ofas_i5k.TPM.scaled.matrix", sep="\t")

write.table(Hvit.TPM.scaled, "C:/Users/cruth/Google Drive/Treehoppers/SRAProject/WGCNA/Hvitripennis/Hvit.90.isoforms.TPM.scaled.matrix", sep="\t")

library(DESeq2)
colData <- read.table("C:/Users/cruth/Google Drive/Treehoppers/SRAProject/WGCNA/Hemiptera_SampleInformation_colData.txt", header=TRUE)
hvCol <- (colData[c(42:62,74:88),])
hvTPM <- as.matrix(Hvit.TPM.scaled)
storage.mode(hvTPM) = "integer"
hv.dds <- DESeqDataSetFromMatrix(hvTPM, colData = na.omit(hvCol), 
                                 design = ~ Tissue + Pool)

hv.dds <- estimateSizeFactors(hv.dds)  ### These steps will take a little bit.... 
hv.dds <- estimateDispersions(hv.dds)
hvScaledNorm <- as.data.frame(assay(hv.dds), normalized = TRUE) ### Normalizes TPM to account for library size
write.table(hvScaledNorm, "Hvit_Scaled_SizeNormed_Integer_TPM.matrix", sep="\t")


ofTPM <- as.matrix(Ofas.TPM.scaled)
ofCol <- colData[c(63:73),]
storage.mode(ofTPM) = "integer"
of.dds <- DESeqDataSetFromMatrix(ofTPM, colData = ofCol, design = ~ Tissue + Pool)
of.dds <- estimateSizeFactors(of.dds)
of.dds <- estimateDispersions(of.dds)
ofScaledNorm <- as.data.frame(assay(of.dds), normalized = TRUE)
write.table(ofScaledNorm, "Ofas_Scaled_SizeNormed_Integer_TPM.matrix", sep="\t")


ECid <- Ecar.TPM[,1]
Ecar.TPM <- Ecar.TPM[,-1]
rownames(Ecar.TPM) <- ECid
ec.dds <- as.matrix(Ecar.TPM)
ecCol <- colData[1:41,]
storage.mode(ec.dds) = "integer"

ec.dds <- DESeqDataSetFromMatrix(ec.dds, colData = ecCol,
                                 design = ~ Tissue + Pool)

ec.dds <- estimateSizeFactors(ec.dds)
ec.dds <- estimateDispersions(ec.dds)
ecNorm <- as.data.frame(assay(ec.dds), normalized=TRUE)
write.table(ecNorm, "Ecar_SizeNormed_Integer_TPM.matrix", sep="\t")


```

That's it. Now we have

1) Scaled the larger transcriptomes (Hvit, Ofas) to the smallest (Ecar)
2) Used DESeq2 to account for different sequencing depths within species
3) Wrote new tables of scaled and size-normalized TPM

The next step in my pipeline is filtering these matrices for single-copy orthologs. 

### R session information
```{r}
sessionInfo()
```
