### Cera Fisher 2019
### This code makes use of the heatmap.3 function by Obi Griffith (https://github.com/obigriffith/biostar-tutorials)
### as modified by Brian Haas of the Broad Institute for the Trinity software suite (https://github.com/trinityrnaseq/trinityrnaseq/blob/devel/Analysis/DifferentialExpression/R/heatmap.3.R). 
### The original code was made available under the GPL, and the Trinity software suite is licensed by 
### the Broad Institute under the license available here: https://github.com/trinityrnaseq/trinityrnaseq/blob/devel/LICENSE.txt
### The code below is my own, and licensed under the MIT License. Use what you like. 


library(DESeq2)
library(dplyr) ## Use dplyr for working with data frames 
options(stringsAsFactors = FALSE)
ECdata <- read.table("ECEF.genes.counts.matrix.collapsed.wing.isos.txt", sep="\t", header=T, row.names = "X")
data <- ECdata
data$id <- rownames(data)
col_ordering = c(1,9,17,
                 2,10,18,
                 3,11,19,
                 4,12,20,
                 5,13,21,
                 6,14,22,
                 7,15,23,
                 8,16,24)

Ecar.gene.orthos <- read.table("CandidateWingRelatedGenes_OrthoTranscriptIDs.txt", sep="\t", header=TRUE)
Ecar.final.id <- Ecar.gene.orthos$Entylia     
# A nice clean regular expression; \\d+ means at one or more digits. We're just stripping off any isoform designators.
Ecar.final.id <- gsub("_i", "", Ecar.final.id)
Ecar.final.id <- gsub(" ", "", Ecar.final.id) # or annoying little spaces generated by Excel
Ecar.final.id <- data.frame(Ecar.final.id)
Ecar.final.id <- mutate(Ecar.final.id, gene.name = Ecar.gene.orthos$gene.name) ### I love dplyr
colnames(Ecar.final.id) <- c("Entylia", "gene.name")

rnaseqMatrix = data[,col_ordering]
rnaseqMatrix = round(rnaseqMatrix)
rnaseqMatrix = rnaseqMatrix[rowSums(rnaseqMatrix)>=2,]
conditions = data.frame(conditions=factor(c(rep("ECAbd", 3), 
                                            rep("ECEye", 3), 
                                            rep("ECLeg", 3), 
                                            rep("ECMeso", 3), 
                                            rep("ECOvi", 3), 
                                            rep("ECPro", 3),
                                            rep("ECWing2", 3),
                                            rep("ECWing3", 3))))
rownames(conditions) = colnames(rnaseqMatrix)
ddsFullCountTable <- DESeqDataSetFromMatrix(
    countData = rnaseqMatrix,
    colData = conditions,
    design = ~ conditions)
ECdds = DESeq(ddsFullCountTable)  # Takes a minute to process

vst <- vst(ECdds)
#### Restart from here by loading the RData
#load("Ecar_WingGenes_data.RData")
vst <- vst(ECdds)
vst <- assay(vst)
res = results(ECdds)
counts.dds <- data.frame(counts(ECdds, normalized=TRUE))

baseMeanAbd <- rowMeans(vst[,c(1,2,3)])
baseMeanEye <- rowMeans(vst[,c(4,5,6)])
baseMeanLeg <- rowMeans(vst[,c(7,8,9)])
baseMeanMeso <- rowMeans(vst[,c(10,11,12)])
baseMeanOvi <- rowMeans(vst[,c(13,14,15)])
baseMeanPro <- rowMeans(vst[,c(16,17,18)])
baseMeanWing2 <- rowMeans(vst[,c(19,20,21)])
baseMeanWing3 <- rowMeans(vst[,c(22,23,24)])

res = cbind(baseMeanAbd, baseMeanEye, baseMeanLeg, baseMeanMeso, baseMeanOvi, baseMeanPro, baseMeanWing2, baseMeanWing3, as.data.frame(res))
res = cbind(id=rownames(res), as.data.frame(res))
res$padj[is.na(res$padj)]  <- 1

### write commands are commented out; uncomment if you want to write data to your machine.
#write.table(res, "Ecar_vst_means.tab",sep = "\t", quote = FALSE, row.names = res$id)
wingRes <- res %>% filter(id %in% Ecar.final.id$Entylia)
#write.table(wingRes, "Ecar_winggenes_vst_means.tab", sep="\t", quote=FALSE, row.names=wingRes$id)

#### heatmap.3.R file from github
source("https://raw.githubusercontent.com/trinityrnaseq/trinityrnaseq/devel/Analysis/DifferentialExpression/R/heatmap.3.R")

res2 <- wingRes[match(Ecar.final.id$Entylia, wingRes$id), ]
colnames(res2)[1] <- "Entylia"
gene.order <- inner_join(res2, Ecar.gene.orthos, by = "Entylia")

## Now "gene.order" has my data in the right order, so I can subset the matrix of means out of it for 
## Entylia:
data <- as.matrix(gene.order[,c(8,9,7,2,5,4,6,3)])

gene.names <- gene.order$gene.name
rownames(data) <- gene.names

myheatcol = colorpanel(100, 'white','palegoldenrod','seagreen4')
column_annotation <- c("violet", "purple", "cyan", "darkblue", "blue", "darkgreen", "lightgreen", "red")
column_annotation <- as.matrix(t(column_annotation))


heatmap.3(data, dendrogram = "none", cexCol = .8, cexRow = .8, Rowv = FALSE, Colv = FALSE, 
          col = myheatcol, main="E.carinata VST(counts)", ColSideColors = column_annotation) 


#pdf("ECWingGenes.vstCounts.Wing2Order-Alphabetical.pdf")
#heatmap.3(data, dendrogram = "none", cexCol = .8, cexRow = .8, Rowv = FALSE, Colv = FALSE, 
#          col = myheatcol, main="E.carinata VST(counts)", ColSideColors = column_annotation) 
#dev.off()


#~~~~~~~~ Uncomment these lines to save the data for reloading and rerunning ~~~~~~~~~~~#
# ECData <- data
# ECdds <- dds
# ECheatcol <- myheatcol
#save(Ecar.final.id, ECheatcol, ECdds, ECData, Ecar.gene.orthos, file="Ecar_WingGenes_data.RData")
